### PR TITLE
DEP: deprecate np.set_numeric_ops and friends

### PR DESCRIPTION
**Pull request by mattip**
The third part of [NEP 15](http://www.numpy.org/neps/nep-0015-merge-multiarray-umath.html#proposed-changes). Deprecate `np.set_numeric_ops`, `PyArray_SetNumericOps`, `PyArray_GetNumericOps`.

The second part, "we should use some new, private API to set up `ndarray.__add__` and friends" is still TBD.

### CODE REVIEW

File: numpy/core/_add_newdocs.py
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]:
```
.. deprecated:: 1.16

    Use :c:func:`PyUFunc_ReplaceLoopBySignature` to change ufunc behaviour.
```

Comment (mattip) [reply numpy/core/_add_newdocs.py]: fixed

**Comment by mattip** [comment]
Codecov failures since newly deprecated `PyArray_GetNumericOps` is not tested.

File: numpy/core/src/multiarray/number.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: nit: weird indent on the line above

File: numpy/core/src/multiarray/number.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Not sure there's much point in this - an exception is already thrown here anyway

File: numpy/core/src/multiarray/number.c
Comment (mattip) [reply numpy/core/src/multiarray/number.c]: fixed

File: numpy/core/src/multiarray/number.c
Comment (mattip) [reply numpy/core/src/multiarray/number.c]: reverted (removed)

**Comment by mattip** [comment]
I think codecov is comparing this pr to some other base changeset. The [diff](https://codecov.io/gh/numpy/numpy/compare/e796449b8dea5f11a959d5664b9a2669add26eed...53613cff0a83508bab5e81019f448fc499056894/changes) is showing changes I did not make to `distutils` and `random`.

**Comment by tylerjereddy** [comment]
Your link is for "changes," not the [actual diff](https://codecov.io/gh/numpy/numpy/compare/e796449b8dea5f11a959d5664b9a2669add26eed...53613cff0a83508bab5e81019f448fc499056894/diff), which correctly shows which lines in this PR are not covered. I tend to ignore the "changes" tab, as previously discussed elsewhere.

**Comment by mattip** [comment]
rebased and deprecation test added to help code coverage

**Comment by mattip** [comment]
Closed/reopened to restart crashed appveyor tests

File: numpy/core/src/multiarray/multiarraymodule.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: Doesn't this mean that you're actually _not_ deprecating the python-side APIs, only the C ones?

Comment (mattip) [reply numpy/core/src/multiarray/multiarraymodule.c]: Since this function is mapped to python-level `set_numeric_ops`, I only want the `DeprecationWarning` from `PyArray_SetNumericOps`, three lines down. There is no python-level `get_numeric_ops`.

File: numpy/core/tests/test_deprecations.py
Comment (mattip) [reply numpy/core/tests/test_deprecations.py]: ok

**Comment by mattip** [comment]
Should squash before merging

**Review by eric-wieser — DISMISSED** [review]

**Comment by mattip** [comment]
squashed to a single commit

File: numpy/core/_add_newdocs.py
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: Perhaps we should also suggest creating a subclass with an overloaded `__array_ufunc__` here, for python-only users who were replacing operators.

Comment (mattip) [reply numpy/core/_add_newdocs.py]: done

**Comment by eric-wieser** [comment]
This closes #10141, right?

**Comment by eric-wieser** [comment]
It doesn't look like you added a deprecation notice to [the C api docs](https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.array.html#c.PyArray_SetNumericOps)

**Comment by mattip** [comment]
C API deprecation added. Also added a commit to remove mention of the deprecated `set_numeric_ops` which closes #10141

**Comment by mattip** [comment]
ping

**Comment by mattip** [comment]
rebased to fix release note merge confilicts

**Comment by mattip** [comment]
ping

File: doc/release/1.16.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.16.0-notes.rst]: Rebase seems to have stripped a newline

**Comment by eric-wieser** [comment]
I feel like we have a compile-time mechanism for deprecating C Apis?

At a guess, perhaps @charris or @njsmith  would know more.

File: doc/release/1.16.0-notes.rst
Comment (mattip) [reply doc/release/1.16.0-notes.rst]: fixed, and found a typo

**Comment by mattip** [comment]
@eric-wieser indeed I did not increment the NPY_API_VERSION for the deprecation. Fixed.

File: numpy/core/src/multiarray/number.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Should be wrapped in
```
#if !defined(NPY_NO_DEPRECATED_API) || \
    (NPY_NO_DEPRECATED_API < NPY_1_16_API_VERSION)

#endif
```

Comment (mattip) [reply numpy/core/src/multiarray/number.c]: That is only for headers, to allow access to old API macros. The C files always start with this define, which will defeat the check you suggest
[`#define NPY_NO_DEPRECATED_API NPY_API_VERSION`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/src/multiarray/number.c#L6). `NPY_API_VERSION` is hard-coded from `C_API_VERSION` in [`core/setup_common.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup_common.py#L44) , see [`core/setup.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup.py#L561) where the magic happens.

We don't really have a way to actually remove or disable deprecated API functions, since the convention is that the PyArray_API structure built in `__multiarray_api.h` by [`core/code_generators/generate_array_api.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/code_generators/generate_numpy_api.py) from the list in [`core.code_generators/numpy_api.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/code_generators/numpy_api.py#L32) can only grow, never shrink, to allow backward compatibility.

This is somewhat talked about in the [documentation](http://www.numpy.org/devdocs/reference/c-api.array.html?highlight=pyarray_api#miscellaneous) from the user's perspective and in the ['VERSIONING SUPPORT'](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/setup_common.py#L12) docstring in `setup_common.py`, but if you think of hook where I can hang more documentation let me know.

Comment (mattip) [reply numpy/core/src/multiarray/number.c]: We could eventually turn the warning into an exception. Not sure if I understand which "C api macro" you mean, there are so many.

At some point we could remove things from the `PyArray_API` structure by changing our major version, which would infer no backward API compatibility.

Comment (eric-wieser) [reply numpy/core/src/multiarray/number.c]: I mean this macro:
```
#define PyArray_GetNumericOps \
        (*(PyObject * (*)(void)) \
         PyArray_API[41])
```

Perhaps we should leave this as an open question, and just take this PR as it currently stands

**Comment by charris** [comment]
Not sure that we should increment the API version, the API hash remains the same and nothing has been added or removed, nor do we remove deprecated API functions -- there are a couple. So the API hasn't changed, we just recommend that some interface functions should not be used. If we want to disable the function, I don't think we have a policy for that.

**Comment by eric-wieser** [comment]
> nor do we remove deprecated API functions

If that's the case, why does `NPY_NO_DEPRECATED_API` exist?

**Comment by mattip** [comment]
Those who wish to access `PyArrayObject` as per pre-1.7-NumPy can set it to NPY_1_6_API_VERSION and then `PyArrayObject` is exposed as a full C struct.

**Comment by charris** [comment]
> Those who wish to access PyArrayObject as per pre-1.7-NumPy

Yep, there was a move to hide the structure internals behind functions so that we would be freed up to change the structures in the future, but that is incomplete and has been stalled for some time.

**Comment by mattip** [comment]
Note PR #11175  changed the size of `PyUFuncObject`. Since that is a publically-facing API we [incremented `NPY_API_VERSION`](https://github.com/numpy/numpy/pull/11175#issuecomment-427286265) even though the hash did not change.

**Comment by charris** [comment]
But that change is not backwards compatible, i.e., code that uses the new attributes will not run on older NumPy. Is that the case for code compiled against 1.16?

**Comment by mattip** [comment]
The only difference in this PR is the deprecation warning that will be emitted on 1.16 if the deprecated C functions are used. Does that not justify incrementing the `NPY_API_VERSION`?

**Comment by eric-wieser** [comment]
> Those who wish to access PyArrayObject as per pre-1.7-NumPy can set it to NPY_1_6_API_VERSION and then PyArrayObject is exposed as a full C struct.

Conversely, I think that users should be able to opt into no longer having `PyArray_GetNumericOps ` available by defining it to `NPY_1_16_API_VERSION` - which would indeed require an api version increment.

**Comment by njsmith** [comment]
> If that's the case, why does NPY_NO_DEPRECATED_API exist?

Honestly, as far as I've ever been able to figure out, there is no good reason for it to exist.

> Yep, there was a move to hide the structure internals behind functions so that we would be freed up to change the structures in the future, but that is incomplete and has been stalled for some time.

It was also done incorrectly... the internals were moved from macros/direct access to inline functions, but inline functions are just like macros: they both expose the structure internals as part of the ABI. And we never had a plan for how to actually remove the old APIs. And I'm not even sure what the point is... I don't think anything on our roadmap actually benefits from hiding the internals of `PyArrayObject`.

Also, I'm pretty sure that all of the C compilers we support have [native support for issuing deprecation warnings](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-deprecated-function-attribute), so I'm not sure why we need to invent our own system?

**Comment by eric-wieser** [comment]
> the internals were moved from macros/direct access to inline functions

Are you sure the inline functions aren't inline only when used within multiarray? I thought I saw something like that somewhere, but perhaps I'm imagining it.

**Comment by mattip** [comment]
The code currently uses functions in the headers which will probably be inlined to be identical to the macros.

> convert them into actual API functions ...

This would incur a performance penalty since the call could no longer be inlined, and would cross a shared-object border. Whether that penalty is significant or not depends on the use case. We *could* adopt a policy where the first call caches a C-compatible representation of the hidden, possibly non-C-representable field, and functions like `reshape` or `resize` invalidate the cache, but as you say "I don't think anything on our roadmap actually benefits from hiding the internals of `PyArrayObject`" 

Besides the 1.7 attribute hiding API changes, we used NPY_API_VERSION for the `RESOLVEWRITEBACKIFCOPY` deprecations, and now for these changes.

Also, PR #12235 fixes the definition of the NPY_1_XX_API_VERSION macros, and is waiting for review

Perhaps we could merge this, and if a change to the API versioning policy is needed open a new issue for that.

**Comment by stefanv** [comment]
This is probably good to go?  It looks fairly uncontroversial.

File: numpy/core/_add_newdocs.py
Comment (charris) [reply numpy/core/_add_newdocs.py]: This reads badly. Maybe
```
        For the general case, use :c:func:`PyUFunc_ReplaceLoopBySignature`.
        For ndarray subclasses, define the ``__array_ufunc__`` method and
        override the relevant ufunc.
```

Comment (charris) [reply numpy/core/_add_newdocs.py]: Ping.

**Comment by mattip** [comment]
reworded documentation

**Comment by charris** [comment]
Thanks Matti.

**Pull request merged**
